src/drivers/snapdragon_pwm_out/snapdragon_pwm_out.cpp

```diff
@@ -0,0 +1,608 @@
+static struct ::dspal_pwm_ioctl_update_buffer *_update_buffer;
+static struct ::dspal_pwm *_pwm;
+
+
+// ToDo: check frequency required
+static const int FREQUENCY_PWM = 400;
+
+
+/* filename: /dev/fs is mapped to /usr/share/data/adsp */
+static const char *MIXER_FILENAME = "/dev/fs/quad_x.main.mix";
+
+// subscriptions
+int	_controls_subs[actuator_controls_s::NUM_ACTUATOR_CONTROL_GROUPS];
+int	_armed_sub;
+int _fd = -1;
+
+// publications
+orb_advert_t    _outputs_pub = nullptr;
+
+// topic structures
+actuator_controls_s _controls[actuator_controls_s::NUM_ACTUATOR_CONTROL_GROUPS];
+orb_id_t _controls_topics[actuator_controls_s::NUM_ACTUATOR_CONTROL_GROUPS];
+actuator_outputs_s  _outputs;
+actuator_armed_s    _armed;
+
+// polling
+uint8_t _poll_fds_num = 0;
+px4_pollfd_struct_t _poll_fds[actuator_controls_s::NUM_ACTUATOR_CONTROL_GROUPS];
+
+// control group
+uint32_t	_groups_required = 0;
+
+// limit for pwm
+pwm_limit_t     _pwm_limit;
+
+// esc parameters
+int32_t _pwm_disarmed;
+int32_t _pwm_min;
+int32_t _pwm_max;
+
+MultirotorMixer *_mixer = nullptr;
+
+
+/*
+ * forward declaration
+ */
+
+void usage();
+
+void start();
+
+void stop();
+
+int pwm_initialize(const char *device);
+
+void pwm_deinitialize();
+
+void send_outputs_pwm(const uint16_t *pwm);
+
+void task_main_trampoline(int argc, char *argv[]);
+
+void subscribe();
+
+void task_main(int argc, char *argv[]);
+
+int initialize_mixer(const char *mixer_filename);
+
+int mixer_control_callback(uintptr_t handle, uint8_t control_group, uint8_t control_index, float &input);
+
+
+
+/*
+ * functions
+ */
+
+int mixer_control_callback(uintptr_t handle,
+			   uint8_t control_group,
+			   uint8_t control_index,
+			   float &input)
+{
+	const actuator_controls_s *controls = (actuator_controls_s *)handle;
+
+	input = controls[control_group].control[control_index];
+
+	return 0;
+}
+
+
+int initialize_mixer(const char *mixer_filename)
+{
+
+	char buf[2048];
+	size_t buflen = sizeof(buf);
+	PX4_INFO("Trying to initialize mixer from config file %s", mixer_filename);
+	int fd_load = ::open(mixer_filename, O_RDONLY);
+	if (fd_load != -1) {
+		int nRead = ::read(fd_load, buf, buflen);
+		close(fd_load);
+
+		if (nRead > 0) {
+			_mixer = MultirotorMixer::from_text(mixer_control_callback, (uintptr_t)&_controls, buf, buflen);
+
+			if (_mixer != nullptr) {
+				PX4_INFO("Successfully initialized mixer from config file");
+				return 0;
+
+			} else {
+				PX4_ERR("Unable to parse from mixer config file");
+				return -1;
+			}
+
+		} else {
+			PX4_WARN("Unable to read from mixer config file");
+			return -2;
+		}
+
+	} else {
+		PX4_WARN("No mixer config file found, using default mixer.");
+
+		/* Mixer file loading failed, fall back to default mixer configuration for
+		* QUAD_X airframe. */
+		float roll_scale = 1;
+		float pitch_scale = 1;
+		float yaw_scale = 1;
+		float deadband = 0;
+
+		_mixer = new MultirotorMixer(mixer_control_callback, (uintptr_t)&_controls,
```

Comment:
Since there is a `new` here, there needs to be a `delete` in `pwm_deinitialize`.